tools/clang/lib/CodeGen/CGCall.cpp

```diff
@@ -1902,7 +1904,7 @@ void CodeGenFunction::EmitFunctionProlog(const CGFunctionInfo &FI,
     case ABIArgInfo::Extend:
     case ABIArgInfo::Direct: {
       // HLSL Change Begins
-      if (hlsl::IsHLSLMatType(Ty)) {
+      if (hlsl::IsHLSLMatType(Ty) || hlsl::IsHLSLHitObjectType(Ty)) {
```

Comment:
This special handling seems odd. Why is this needed?

Reply:
This is necessary because otherwise call lowering will flatten the HitObject struct.

Reply:
See my other comment, but we should not be matching the pattern set by matrices because it is an endless source of bugs.

Reply:
Removed in 358942da6 . HitObject is now an aggregate.